### PR TITLE
tests: Use libvmi instead of 'NewRandomVMIWithEphemeralDiskAndUserdataNetworkData'

### PR DESCRIPTION
**Pull request by ormergi**
### What this PR does
Use libvmi directly at the caller instead of `NewRandomVMIWithEphemeralDiskAndUserdataNetworkData`.

Before this PR:

After this PR:

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
- Since LaunchVMI() returned value is not used, the VMI instance in the
test context never get updated with the namespace causing failures.
It worked because 'NewRandomVMIWithEphemeralDiskAndUserdataNetworkData'
set the VMI with a namespace, so the instance in the test context had
the namespace.
To address this assign the the returned value from LaunchVMI().

- Regarding the test 'should have cloud-init network-config from k8s secret':
NewCirros() set cloudInit by default (to prevent it to hang at boot time).
Since the test asserts cloud-init user-data (and encoded use-data) is empty,
override cloudinit user-data that is set by NewCirros() as before these commit
changes.
Since cloud-init network-data is set the guest doesn't hang at boot time.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/libvmi/cloudinit.go
Comment (orelmisan) [inline tests/libvmi/cloudinit.go:70]: nit: `encodedData` is used only once and could be inlined:
```suggestion
		volume.CloudInitNoCloud.NetworkDataBase64 = base64.StdEncoding.EncodeToString([]byte(data))
```

File: tests/vmi_cloudinit_test.go
Comment (orelmisan) [inline tests/vmi_cloudinit_test.go:355]: Is it required to explicitly define a network and an interface in this test?
AFAIU, the test only checks the network config file.
Same below.

File: tests/vmi_cloudinit_test.go
Comment (dankenigsberg) [inline tests/vmi_cloudinit_test.go:393]: As you mention in the commit message, NewCirros sets cloudinit once; here it is erased; and third time sets it to a `NetworkDataSecretRef`. It can be nicer with a helper function `withCloudInitNoCloudNetworkDataSecret(secretName)`

File: tests/vmi_cloudinit_test.go
Comment (ormergi) [reply tests/vmi_cloudinit_test.go:355]: `NewRandomVMIWithEphemeralDiskAndUserdataNetworkData` returned VMI with masquerade interface, therefore I followed.
If we set no interface, bridge interface will be added by default and it should work as expected.
Nevertheless, I think its better specifying the interface for better visibility, it may look odd for the next reviewer that thetest configure networking but no interface is specified.
Also, in case this default behavior miraculously change, this test would still be valid.

File: tests/libvmi/cloudinit.go
Comment (ormergi) [reply tests/libvmi/cloudinit.go:72]: Done

**Comment by ormergi** [comment]
@dankenigsberg @orelmisan please take another look :slightly_smiling_face:

File: tests/libvmi/cloudinit.go
Comment (dankenigsberg) [reply tests/libvmi/cloudinit.go:72]: sorry, I don't notice a change here - you still do not erase the non-encoded field.

File: tests/libvmi/cloudinit.go
Comment (dankenigsberg) [inline tests/libvmi/cloudinit.go:25]: must we add this dependency to libvmi? maybe you can keep `WithCloudInitNoCloudEncodedNetworkData` out for now? I'm not sure myself.

p.s you use a very rare name for the imported package.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/libvmi/cloudinit.go
Comment (ormergi) [reply tests/libvmi/cloudinit.go:72]: My bad I pushed the wrong changes, PTAL

Comment (ormergi) [reply tests/libvmi/cloudinit.go:72]: Done

File: tests/libvmi/cloudinit.go
Comment (ormergi) [reply tests/libvmi/cloudinit.go:66]: Done

File: tests/libvmi/cloudinit.go
Comment (ormergi) [reply tests/libvmi/cloudinit.go:25]: `WithCloudInitNoCloudNetworkDataSecret` introduce this dependency.
Unfortunately its not possible to avoid because the Volume API depndes on it (`volume.CloudInitNoCloud.NetworkDataSecretRef` is type of `k8s.io/api/core/v1`).

Comment (dankenigsberg) [reply tests/libvmi/cloudinit.go:25]: I know... Hence I suggest to keep this function out of this package. It has one users, so it can be defined there. As I said, I'm not 100% sure myself.

File: tests/libvmi/cloudinit.go
Comment (EdDev) [inline tests/libvmi/cloudinit.go:83]: What about passing in `NetworkDataSecretRef` value as the input?.

File: tests/libvmi/cloudinit.go
Comment (EdDev) [reply tests/libvmi/cloudinit.go:25]: The dependency already exists transitively, as `kubevirt.io/api/core/v1` depends on it.

**Comment by enp0s3** [comment]
/approve

Despite the dependency concern if we move them to the cloud_init test file we will have to copy the helper functions of libvmi cloud-init there.

File: tests/libvmi/cloudinit.go
Comment (EdDev) [reply tests/libvmi/cloudinit.go:25]: > @EdDev please use your `lgtm` powers, then.

There are other opened comments  and the linter is failing.

File: tests/libvmi/cloudinit.go
Comment (ormergi) [reply tests/libvmi/cloudinit.go:83]: It wont do because it would be mentioned in the function signature and introduce dependency on `k8s.io/api/core/v1`

Comment (EdDev) [reply tests/libvmi/cloudinit.go:82]: It is explicit about what you change with a type check. Not related to the dependency. But if you prefer it like this I do not mind much.

Comment (ormergi) [reply tests/libvmi/cloudinit.go:82]: I tried to make it easy to work with in tests, the test shouldn't care for this wrapping object(`LocalObjectRefrence`1), just the secret object name it that it creates.

Yeah, I prefer to leave it as is.

**Comment by ormergi** [comment]
rebased

File: tests/vmi_cloudinit_test.go
Comment (ormergi) [reply tests/vmi_cloudinit_test.go:410]: Posted #11545

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [inline pkg/libvmi/cloudinit.go:82]: Is `NetworkDataSecretRef` mutually exclusive with `NetworkData` and `NetworkDataBased64`? If so, I think you should set it to `nil` on every `WithCloudInit*`

Comment (ormergi) [reply pkg/libvmi/cloudinit.go:82]: I think it may contradict with other cloud-init settings yes, but I need to test it so I prefer to address it in a follow up PR.

Comment (dankenigsberg) [reply pkg/libvmi/cloudinit.go:82]: /hold

Please reconsider. If my hunch is correct, this new function is introducing an awkward semantic.
Besides, since this changes the API of `libvmi`, I think it would be nice if its owner @EdDev takes another look.

Comment (ormergi) [reply pkg/libvmi/cloudinit.go:82]: This PR already address more then its intent following review comments, I prefer to addressed additional stuff on a follow up PR.
If you insistent on having  these changes, I can move all changes that are not related to this PR initial intent to another PR - second and third commits - along with what you just mention.

Comment (ormergi) [reply pkg/libvmi/cloudinit.go:82]: Nevertheless, this PR doesn't introduce the issue you mention, it just surface it.
The same issue exist also for the user-data, the API enable referring to secret (as with network-data) but non of `WithCloudInit*` overrides it.

Comment (EdDev) [reply pkg/libvmi/cloudinit.go:82]: > Is `NetworkDataSecretRef` mutually exclusive with `NetworkData` and `NetworkDataBased64`? If so, I think you should set it to `nil` on every `WithCloudInit*`

If we follow the logic-less build pattern, I think we should leave it up to the caller to correctly build the spec. As a tester, the caller may be interested in building a spec that is indeed colliding in order to check the validation or whatever outcome this ends with.

Does this make sense?

On the other hand, this pattern of initializing the other fields has been already taken in the other options (when setting the encoded and non-encoded versions). Even here the two are initialized.
Therefore, it does make sense to perform the initialization.

It should be fixed one way or the other and I agree it cannot be deferred to later.
The simplest option would be to do the initialization, as it continues the existing pattern.
In a follow up we can consider dropping it for all.

Another problem I see is that we assume a hard-coded disk name, which made sense when serving e2e tests only, now it is a bit awkward.
If you have a good idea how to solve it, please let me know. I guess this can also be done later.

File: pkg/libvmi/cloudinit.go
Comment (EdDev) [inline pkg/libvmi/cloudinit.go:77]: I still prefer to pass here a secretRef: `secretRef *k8scorev1.LocalObjectReference`.
But it is not a blocker.

If you keep it like this, perhaps it is better to add `Name` at the end of the function name and mention it in the description.

File: tests/vmi_cloudinit_test.go
Comment (EdDev) [inline tests/vmi_cloudinit_test.go:675]: Can you replace `get` with `lookup`?

File: tests/vmi_cloudinit_test.go
Comment (EdDev) [inline tests/vmi_cloudinit_test.go:443]: I think it will make more sense to check equality with the whole `CloudInitNoCloud` structure. Is it possible?
`Expect(testVolume.CloudInitNoCloud).To(Equal(...))`

**Review by EdDev — CHANGES_REQUESTED** [review]

File: pkg/libvmi/cloudinit.go
Comment (EdDev) [reply pkg/libvmi/cloudinit.go:82]: > Nevertheless, this PR doesn't introduce the issue you mention, it just surface it. The same issue exist also for the user-data, the API enable referring to secret (as with network-data) but non of `WithCloudInit*` overrides it. I suggest addressing it and fix all cloud-init options libvmi functions on different PR.

I do not think this is correct. You have added another options to add data to cloud-init which was not there before. Now, someone may add this data and it is not in sync with the other options.
So I think it does introduce an issue, not only surface it.

This change does initialize the encoded and non-encouded versions, so I think it is obvious that the others should reset the secret field as well.

File: tests/vmi_cloudinit_test.go
Comment (ormergi) [reply tests/vmi_cloudinit_test.go:443]: TBH I am not sure why this assert is necessary at all, this PR doesn't change it.
I prefer to do that on a follow up as other tests follow the same pattern.
It also been said in other thread we should revisit cloud-init tests do some refactoring, including the assertions as you suggest, and remove/consolidate tests.

File: pkg/libvmi/cloudinit.go
Comment (ormergi) [reply pkg/libvmi/cloudinit.go:77]: Done, added `Name` at the end.

File: pkg/libvmi/cloudinit.go
Comment (ormergi) [reply pkg/libvmi/cloudinit.go:82]: Thanks for explaining.
I changed other `libvmi.WithCloudInit..` that set network-data to override the NetworkSecretRef field.
Also fixed the introduced function to override the correct fields.

File: tests/vmi_cloudinit_test.go
Comment (ormergi) [reply tests/vmi_cloudinit_test.go:675]: Done

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [inline pkg/libvmi/cloudinit.go:62]: this line fixes an old bug, consider putting it in its own commit.

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [inline pkg/libvmi/cloudinit.go:63]: I think that this line should be added elsewhere, too. Eg https://github.com/kubevirt/kubevirt/pull/11310/commits/7b2b782f4ed7b5785bdbab3fe8734638e4611aa7#diff-7cd66099e0fe3bc5ccbe619f796d159e11384d71d06147bae0a6ff7c829284f6R51

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [inline pkg/libvmi/cloudinit.go:74]: ditto

File: pkg/libvmi/cloudinit.go
Comment (ormergi) [reply pkg/libvmi/cloudinit.go:63]: The referenced function, 'WithCloudInitNoCloudEncodedUserData', configure user-data, it should not override network-data configuration.

File: pkg/libvmi/cloudinit.go
Comment (ormergi) [reply pkg/libvmi/cloudinit.go:74]: Here ,`WithCloudInitNoCloudEncodedNetworkData`, edit network-data, it should not override user-data configuration.

File: pkg/libvmi/cloudinit.go
Comment (ormergi) [reply pkg/libvmi/cloudinit.go:62]: Done

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [reply pkg/libvmi/cloudinit.go:85]: Thanks!

/hold cancel

**Review by dankenigsberg — APPROVED** [review]

**Review by EdDev — APPROVED** [review]
Thanks.

File: pkg/libvmi/cloudinit.go
Comment (dankenigsberg) [reply pkg/libvmi/cloudinit.go:74]: thanks, my bad.

**Comment by dankenigsberg** [comment]
Compute lane seems to consistently fail. Please review and unhold.

/hold

**Comment by ormergi** [comment]
/unhold 

Last changes should fix the issue

**Review by dankenigsberg — APPROVED** [review]

**Comment by ormergi** [comment]
Rebased

**Review by dankenigsberg — APPROVED** [review]

**Pull request merged**
